File: Data Structures & Algorithms/eating-bananas/submission-0.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int maxPiles = 0;
        for (int pile : piles)
            maxPiles = Math.max(maxPiles, pile);

        int left = 1, right = maxPiles;

        while (left < right) {
            int mid = left + (right - left) / 2;
            long totalHours = 0;

            for (int pile : piles) 
                totalHours += (pile + mid - 1) / mid;

            if (totalHours <= h) 
                right = mid;
            else 
                left = mid + 1;
        }

        return left;
    }
}

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-0.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> freq;
        int left = 0, right = 0, maxSize = 0;

        while (right < s.size()) {
            freq[s[right] - 'a']++;

            while (freq[s[right] - 'a'] > 1)
                freq[s[left++] - 'a']--;

            maxSize = max(maxSize, right - left + 1);
            right++;
        }

        return maxSize;
    }
};
